golang/session: initial implementation and basic test

progrium opened the pull request:
This is the meat of the implementation. Although I'd like better test coverage and comments, I'm particularly concerned about the use of contexts. I don't believe they are used where given or appropriate where given so I'd like to think about them more. 

Also, I don't often centralize errors into predefined variables, but it might be a good idea to here? 

Double check min-max limits ... there are packet sizes and data packet sizes and I'm not sure they're correctly used?

After this there is simply a transport subpackage with various transport protocol dialers and such.

If synchronous review works better, schedule a time with me!

shazow commented on golang/session/util_buffer.go:
```diff
@@ -0,0 +1,93 @@
+// buffer provides a linked list buffer for data exchange
+// between producer and consumer. Theoretically the buffer is
+// of unlimited capacity as it does no allocation of its own.
+type buffer struct {
```
Hm, what's the upside of the linked list approach vs building on top of https://golang.org/pkg/sync/#Pool?

progrium replied:
I was also considering that this could be satisfied by a `sync.Map` but since there are some cases where they are slower and this was originally in the /x/crypto/ssh implementation (where most of this code came from), I figured I should leave it. I didn't think about a Pool but that's even harder to see how it would replace. Thoughts?

shazow replied:
I think what threw me off is I assumed this exists to reduce allocations, which is also the point of sync.Pool, but now that I read closer it's more about providing concurrent read/write access to chunks in a sequential way. Does that sound right?

progrium replied:
Yes I believe so. Instead of locking for one big buffer, a reader can read chunks off while there are no writes/eofs, and writes/eofs can only be done sequentially.

progrium replied:
Oh, sorry ... I was thinking chanList could be replaced by `sync.Map`, I was confused when I tried to make sense of that.

shazow commented on golang/session/channel.go:
```diff
@@ -0,0 +1,251 @@
+// channel is an implementation of the Channel interface that works
+// with the session class.
+type channel struct {
+	ctx context.Context
```
I guess the ctx doesn't do anything at all yet?

progrium replied:
There were two cases I had in mind. 

The first was timeouts. More APIs that have a long running, cancelable bit of work usually take a context for timeouts, but what are reasonable timeouts here? Maybe a session/conn can be canceled? If `net.Conn` doesn't care then I guess we wouldn't here. However, a Channel could have a context because a Channel is not necessarily a connection, it could be a request. So looking at how net/http uses them with request is probably helpful. 

The second was contextual meta-data. Context becomes a a key/value store that is passed down a thread of work that can contain relevant data or objects that come from up the calling code. There is much less consensus on the right way to use this, as you can argue those should be part of the API. However, too many times in my experience it can be used as a way to pass data in that the author didn't think to allow, so it's just a helpful thing to have. But I think that's also important to pass data that is not relevant to the domain of the library, but a domain it might be used. If you use Channel as a request, maybe you want to keep a trace ID with it. 

So I took the approach of being able to add a context to a session, which was passed to a channel, so there is a hierarchy to top level values that might be associated with a session you can have with a channel if you are passing just a channel around. Maybe the channel should have its own context in which the session is in. You could say maybe that should be part of the API, but I think it makes more sense in a context.

So yea.

progrium replied:
After studying net and http packages and reading the article, I've removed contexts since there are other ways to associate values with a channel (wrapping it in a Request that may have its own context setup). The only context that makes sense is for Open, which I've run into situations where I'd like it to timeout, so I added that.

shazow reviewed (commented):
Looks plausible! Definitely very ssh-esque.

Also is auth going to be part of qmux at all, or are we assuming a pre-authed io stream?

One other scenario that's worth keeping in mind is somekind of graceful reload support when possible, without losing state/connections. That's one thing that's particularly difficult with ssh.

mgood reviewed (commented):
I'm still reading, but also curious about the context

progrium commented:
> Also is auth going to be part of qmux at all, or are we assuming a pre-authed io stream?

Auth is out of scope. Either it's handled in the underlying transport (WebSocket, for example), preceding the session (TLS client certs maybe), or in a protocol above this (maybe you're muxing HTTP, or maybe you have an RPC auth mechanism). 

> One other scenario that's worth keeping in mind is somekind of graceful reload support when possible, without losing state/connections. That's one thing that's particularly difficult with ssh.

That's hard in general right? What are some examples?

shazow commented:
> That's hard in general right? What are some examples?

Yea. There's lots of implementations for this in HTTP, but HTTP is fairly stateless so it's less hard. Here's a fairly general TCP implementation: https://github.com/facebookarchive/grace/blob/75cf19382434/gracenet/net.go#L206

This is probably another transport-level thing but it helps when the protocol/API doesn't get in the way.

progrium commented:
@mgood @shazow besides removing stored contexts, and adding context to Open, I've also removed the Listener interface (for now) and dropped the LocalAddr/RemoteAddr methods from Session, since if Session is on a net.Conn, you should be able to type assert as net.Conn to get those. They don't make sense if over stdio for example.

shazow commented:
That does look better.

progrium commented:
I'll take that as a pass for now!
